File: Longest valid Parentheses.py
class Solution:
    def maxLength(self, s):
        stack = [-1]  # base index
        max_len = 0

        for i, ch in enumerate(s):
            if ch == '(':
                stack.append(i)
            else:
                stack.pop()
                if len(stack) == 0:
                    stack.append(i)
                else:
                    max_len = max(max_len, i - stack[-1])
        return max_len

File: Longest valid Parentheses.java
class Solution {
    static int maxLength(String s) {
        // code here
        Stack<Integer> st = new Stack<Integer>();
        st.push(-1);
        int len=0;
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='('){
                st.push(i);
            }
            else{
                if(!st.isEmpty())
                st.pop();
            
            if(!st.isEmpty()){
                len=Math.max(len,i-st.peek());
            }
            else{
                st.push(i);
            }
            }
        }
        return len;
        
    }
}
